userspace/lib/usbip_forward.c

```diff
@@ -259,47 +262,56 @@ swap_iso_descs_endian(char *buf, int num)
 	}
 }
 
-#define OUT_Q_LEN 256
-static long out_q_seqnum_array[OUT_Q_LEN];
+/*
+ * This is a 'usbip_header_basic' cache to hold transfer direction of all
+ * OUT CMD_SUBMIT packets. Cache info is used when RET_SUBMIT packets with the
+ * same sequence number arrive.
+ * The transfer direction, EP address and device ID are not provided in return
+ * RET_SUBMIT packets from Linux, when used as an USBIP server.
+ *
+ * The transfer direction is needed to determine USBIP_RET_SUBMIT packet size
+ * in this example:
+ * The OUT ISOCHRONOUS transfer sends data buffer and its ISO descriptors towards
+ * the device in CMD_SUBMIT packets with 'actual_size' record set to define the
+ * data buffer size. However the return RET_SUBMIT packet of the same OUT transfer
+ * contain only ISO descriptor and the 'actual_size' is set to the sent size value.
+ *
+ * Each cache entry may be written or read many times, however the sequence
+ * number of a cache entry location is kept until current sequence number of
+ * packets increases its value for the number of all cache entries.
+ */
 
-static BOOL
-record_outq_seqnum(unsigned long seqnum)
-{
-	int	i;
+struct usbip_cached_hdr {
+	UINT32 seqnum;
+	// UINT32 devid;
+	UINT32 direction;
+	// UINT32 ep;
+};
 
-	for (i = 0; i < OUT_Q_LEN; i++) {
-		int	found_empty_slot;
+#define HDRS_CACHE_SIZE 1024
+static struct usbip_cached_hdr hdrs_cache[HDRS_CACHE_SIZE];
```

Comment:
:+1: This scheme looks smarter than the old one.

Reply:
@cezanne : thanks, i hope it holds water. Anyway, do you maybe know why the server doesn't set those header fields correctly, so that the cache would not be needed? Is this some security or performance measure?

Reply:
> Anyway, do you maybe know why the server doesn't set those header fields correctly, so that the cache would not be needed? Is this some security or performance measure?

Linux version does not care some reply header fields because a client maintains requests until they are received from the server. Thus, the linux client is able to figure out the direction without a response header. However, a forwarder in usbip-win should cache headers of requests, which is a separate user process apart from a vhci driver. It's a sort of implementation problem?? :cry:

Reply:
Yes, this is unfortunate. It would have been easier to understand, if those fields weren't in the protocol at all - thanks